### PR TITLE
Zonal stats fixes and improvements

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR fixes a bunch of identified issues with the zonal stats code, and greatly speeds up the algorithm in the case of small polygons vs pixel sizes. Lots of new tests too!

### CODE REVIEW

**Comment by nyalldawson** [comment]
@nirvn the zonal histogram algorithm suffers from many of the same bugs - it would be nice to remove the duplicate code here :)

**Comment by nyalldawson** [comment]
@nirvn you should be able to make those functions take a function argument, and pass a lambda function in which handles the calculation of the result

**Comment by nyalldawson** [comment]
@nirvn those methods aren't exposed publically. What I'd suggest doing is making a QgsZonalStatisticsUtils class, with the three methods as static members, and the whole class not exposed to python. Much like how the overlay utils is done.

**Comment by nirvn** [comment]
@nyalldawson , OK, I'll undertake that when this PR is merged.

File: src/analysis/vector/qgszonalstatistics.cpp [hunk 1]
```diff
@@ -439,59 +430,61 @@ void QgsZonalStatistics::statisticsFromMiddlePointTest( const QgsGeometry &poly,
...
     currentY -= cellSizeY;
   }
-  delete block;
 }
 
-bool QgsZonalStatistics::validPixel( float value ) const
+bool QgsZonalStatistics::validPixel( double value ) const
 {
-  return !( value == mInputNodataValue || std::isnan( value ) );
+  return !std::isnan( value );
```
Comment (nirvn) [inline src/analysis/vector/qgszonalstatistics.cpp]: Do we really need this validPixel function here?

Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.cpp]: Do you mean "do we need the check" or "does this have to be a function"?

Comment (nirvn) [reply src/analysis/vector/qgszonalstatistics.cpp]: @nyalldawson , "does this have to be a function?" :smile:

Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.cpp]: Not sure... I'm thinking there's *probably* an issue if non-finite pixel values are present here which should also be caught by the function, but I'm reluctant to add this extra check without knowing that there is actually an issue. So I'd keep it like this to make it easy to possibly refine like this in future....

Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.cpp]: (it's also called in more than one place)

Comment (nirvn) [reply src/analysis/vector/qgszonalstatistics.cpp]: OK

File: src/analysis/vector/qgszonalstatistics.cpp [hunk 2]
```diff
@@ -31,12 +31,37 @@
...
   , mRasterBand( rasterBand )
   , mPolygonLayer( polygonLayer )
   , mAttributePrefix( attributePrefix )
   , mStatistics( stats )
-{}
+{
+  if ( mCellSizeX < 0 )
+  {
+    mCellSizeX = -mCellSizeX;
+  }
```
Comment (pblottiere) [inline src/analysis/vector/qgszonalstatistics.cpp]: Would it be possible to do something like `, mCellSizeX( std::abs( rasterUnitsPerPixelX ) )` instead?

(same question about `mCellSizeY`)

File: src/analysis/vector/qgszonalstatistics.h [hunk 3]
```diff
@@ -106,45 +140,48 @@ class ANALYSIS_EXPORT QgsZonalStatistics
...
-        double sum;
-        double count;
-        float max;
-        float min;
-        QMap< float, int > valueCount;
-        QList< float > values;
+        double sum = 0.0;
+        double count = 0.0;
+        double max = std::numeric_limits<double>::lowest();
+        double min = std::numeric_limits<double>::max();
```
Comment (pblottiere) [inline src/analysis/vector/qgszonalstatistics.h]: I didn't know about initializing double/float members in headers...

Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.h]: It's a C++11 feature, along with the nice delegated constructor I've used here :)

Comment (pblottiere) [reply src/analysis/vector/qgszonalstatistics.h]: Is it something we want to generalize? I mean, I was aware that we wanted to initialize nullptr and boolean in headers. But do we want to initialize all members directly in headers if possible? 

Just to know the rules for the future :).

And yes, delegate constructors rocks :+1:! But last time I tried, it seems to me that I had some issues for compiling on Windows... However, it was some time ago :).

Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.h]: In my view, yes initialise in headers wherever possible. It reduces the chance of uninitialised members being introduced when code is refactored, and reduces duplication (default values explicitly included in all individual constructors). 

I've run the associated clang tidy modernize fixits over the whole code previously, but they haven't caught everything (I think the fixit fails when there's multiple constructors, and it also doesn't like initialisation in the body of constructors (which some old old qgis code does)). So if you see this anywhere, please fix :)

Comment (pblottiere) [reply src/analysis/vector/qgszonalstatistics.h]: OK, thanks for the explanation Nyall! I'll pay attention to these points :)

File: src/analysis/vector/qgszonalstatistics.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/analysis/vector/qgszonalstatistics.cpp]: Fixed

**Pull request merged**